examples/vision/learnable_resizer.py

```diff
@@ -0,0 +1,278 @@
+
+    # Intermediate resizing as a bottleneck.
+    bottleneck = layers.experimental.preprocessing.Resizing(
+        *TARGET_DIM, interpolation=interpolation
+    )(x)
+
+    # Residual passes.
+    for _ in range(num_res_blocks):
+        x = res_block(bottleneck)
+
+    # Projection.
+    x = layers.Conv2D(
+        filters=filters, kernel_size=3, strides=1, padding="same", use_bias=False
+    )(x)
+    x = layers.BatchNormalization()(x)
+
+    # Skip connection
+    x = layers.Add()([bottleneck, x])
+
+    # Final resized image
+    x = layers.Conv2D(filters=3, kernel_size=7, strides=1, padding="same")(x)
+    final_resize = layers.Add()([naive_resize, x])
+
+    return final_resize
+
+
+"""
+## Visualize the outputs of the learnable resizing module
+
+Here, we visualize how the resized images would look like after being passed through the
+random weights of the resizer. 
+"""
+
+sample_images, _ = next(iter(train_ds))
+
+plt.figure(figsize=(10, 10))
+for i, image in enumerate(sample_images[:9]):
+    ax = plt.subplot(3, 3, i + 1)
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    resized_image = learnable_resizer(image[None, ...])
+    plt.imshow(resized_image.numpy().squeeze())
+    plt.axis("off")
+
+"""
+## Model building utility
+"""
+
+
+def get_model():
+    backbone = tf.keras.applications.DenseNet(weights=None, include_top=True, classes=2)
+    backbone.trainable = True
+
+    inputs = layers.Input((INP_DIM[0], INP_DIM[1], 3))
+    x = layers.experimental.preprocessing.Rescaling(scale=1.0 / 255)(inputs)
+    x = learnable_resizer(x)
+    outputs = backbone(x)
+
+    return tf.keras.Model(inputs, outputs)
+
+
+"""
+The structure of the learnable image resizer module allows for flexible integrations with
+different vision models.
+"""
+
+"""
+## Compile and train our model with learnable resizer
+"""
+
+model = get_model()
+model.compile(
+    loss=keras.losses.CategoricalCrossentropy(label_smoothing=0.1),
+    optimizer="sgd",
+    metrics=["accuracy"],
+)
+model.fit(train_ds, validation_data=validation_ds, epochs=EPOCHS)
+
+"""
+## Visualize the outputs of the trained visualizer
+"""
+
+learned_resizer = tf.keras.Model(model.input, model.layers[-2].output)
+
+plt.figure(figsize=(10, 10))
+for i, image in enumerate(sample_images[:9]):
+    ax = plt.subplot(3, 3, i + 1)
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    resized_image = learned_resizer(image[None, ...])
+    plt.imshow(resized_image.numpy().squeeze())
+    plt.axis("off")
+
+"""
+The plot shows that the visuals of the images have improved with training. The following
+table shows the benefits of using the resizing module in comparison to using the bilinear 
+interpolation:
+
+|           Model           	| Number of  parameters (Million) 	| Top-1 accuracy 	|
+|:-------------------------:	|:-------------------------------:	|:--------------:	|
+|   With learnable resizer  	|             7.051717            	|      52.02     	|
```

Comment:
The resizer is a residual convnet. Isn't the improved accuracy simply a consequence of adding more layers (representational power) to the model, compared to the baseline? Maybe this is explained in the paper (didn't have time to read it). How do we argue the scientific interest of these results?

Reply:
This is a great question. So, to show it empirically here's what the authors have done:

* Take a pre-trained model trained some size, say (224 x 224). 
* Now, first, use it to infer predictions on images resized to a lower resolution. Record the performance. 
* For the second experiment, plug in the resizer module at the top of the pre-trained model and warm-start the training. Record performance. 

Now, they argue that using the second option is better because it helps the model learn how to adjust the representations better with respect to the given resolution. I agree this might be because of an increase in the number of parameters (which I have mentioned too) more experiments would have been better to establish this claim. Experiments such as analyzing the channel-similarity index and cross-channel interaction, visualizing if the focus of the network gets better with the resizer (Grad-CAM could be used to do this), and so on.

Reply:
That makes sense! Is this explained in the example sufficiently clearly?

Reply:
I can add these in the example briefly if you'd like.

Reply:
I added the following to clarify:

```
Now, a question worth asking here is -  _isn't the improved accuracy simply a consequence
of adding more layers (the resizer is a mini network after all) to the model, compared to
the baseline?_

To show that it is not the case, the authors conduct the following experiment:

* Take a pre-trained model trained some size, say (224 x 224).

* Now, first, use it to infer predictions on images resized to a lower resolution. Record
the performance.

* For the second experiment, plug in the resizer module at the top of the pre-trained
model and warm-start the training. Record the performance.

Now, the authors argue that using the second option is better because it helps the model
learn how to adjust the representations better with respect to the given resolution.
Since the results purely empirical, a few more experiments such as analyzing the
cross-channel interaction would have been even better. It is worth noting that elements
like [Squeeze and Excitation (SE) blocks](https://arxiv.org/abs/1709.01507), [Global Context (GC) blocks](https://arxiv.org/pdf/1904.11492) also add a few
parameters to an existing network but they are known to help a network process
information in systematic ways to improve the overall performance. 
```